kernel/kernel-api/src/main/java/io/delta/kernel/internal/TransactionBuilderImpl.java

```diff
@@ -0,0 +1,185 @@
+
+    public TransactionBuilderImpl(TableImpl table, String engineInfo, Operation operation) {
+        this.table = table;
+        this.engineInfo = engineInfo;
+        this.operation = operation;
+    }
+
+    @Override
+    public TransactionBuilder withSchema(TableClient tableClient, StructType newSchema) {
+        this.schema = Optional.of(newSchema); // will be verified as part of the build() call
+        return this;
+    }
+
+    @Override
+    public TransactionBuilder withPartitionColumns(
+            TableClient tableClient,
+            List<String> partitionColumns) {
+        if (!partitionColumns.isEmpty()) {
+            this.partitionColumns = Optional.of(partitionColumns);
+        }
+        return this;
+    }
+
+    @Override
+    public Transaction build(TableClient tableClient) {
+        SnapshotImpl snapshot;
+        try {
+            snapshot = (SnapshotImpl) table.getLatestSnapshot(tableClient);
+        } catch (TableNotFoundException tblf) {
+            logger.info(
+                    "Table {} doesn't exist yet. Trying to create a new table.",
+                    table.getPath(tableClient));
+            schema.orElseThrow(() -> new IllegalArgumentException(
+                    "Table doesn't exist yet. Must provide a new schema."));
+            // Table doesn't exist yet. Create an initial snapshot with the new schema.
+            Metadata metadata = getInitialMetadata();
+            Protocol protocol = getInitialProtocol();
+            LogReplay logReplay = getEmptyLogReplay(tableClient, metadata, protocol);
+            snapshot = new InitialSnapshot(
+                    table.getDataPath(),
+                    logReplay,
+                    metadata,
+                    protocol);
+        }
+
+        validate(tableClient, snapshot);
+
+        return new TransactionImpl(
+                table.getDataPath(),
+                table.getLogPath(),
+                snapshot,
+                engineInfo,
+                operation,
+                schema,
+                partitionColumns,
+                snapshot.getProtocol(),
+                snapshot.getMetadata());
+    }
+
+    /**
+     * Validate the given parameters for the transaction.
+     */
+    private void validate(TableClient tableClient, SnapshotImpl snapshot) {
+        // Validate the table has no features that Kernel doesn't yet support writing into it.
+        TableFeatures.validateWriteSupportedTable(
+                snapshot.getProtocol(),
+                snapshot.getMetadata(),
+                snapshot.getMetadata().getSchema());
+
+        long currentSnapshotVersion = snapshot.getVersion(tableClient);
+        if (currentSnapshotVersion >= 0 &&
+                (partitionColumns.isPresent() || schema.isPresent())) {
+            throw new IllegalArgumentException(
+                    "Partition columns can only be set on a table that doesn't exist yet.");
+        }
+
+        if (currentSnapshotVersion < 0) {
```

Comment:
What if the table already exists and user still provides a new schema?

Reply:
We throw exception in the `build`